src/compile/compile.ts

```diff
@@ -65,23 +69,43 @@ function getTopLevelProperties(topLevelSpec: TopLevel<any>, config: Config) {
  * Note: this couldn't be `model.assemble()` since the top-level model
  * needs some special treatment to generate top-level properties.
  */
-function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties) {
+function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties & LayoutSizeMixins) {
   // TODO: change type to become VgSpec
 
   // Config with Vega-Lite only config removed.
   const vgConfig = model.config ? stripAndRedirectConfig(model.config) : undefined;
 
   // autoResize has to be put under autosize
-  const {autoResize, ...topLevelProps} = topLevelProperties;
+  const {autosize, ...topLevelProps} = topLevelProperties;
 
   const title = model.assembleTitle();
   const style = model.assembleGroupStyle();
 
+  let layoutSignals = model.assembleLayoutSignals();
+
+  // move width and height signals with values to top level
+  layoutSignals = layoutSignals.filter(signal => {
+    if ((signal.name === 'width' || signal.name === 'height') && signal.value !== undefined) {
+      topLevelProps[signal.name] = signal.value;
+      return false;
+    }
+    return true;
+  });
+
+  if (autosize === 'fit' || (autosize && isAutosize(autosize) && autosize.type === 'fit')) {
+    if (!topLevelProps.width || !topLevelProps.height) {
+      log.warn(log.message.FIT_DATA_DRIVEN);
+    }
+
+    if (!(model instanceof UnitModel) && !(model instanceof LayerModel)) {
+      log.warn(log.message.FIT_NON_SINGLE);
+    }
+  }
+
   const output = {
     $schema: 'https://vega.github.io/schema/vega/v3.0.json',
     ...(model.description ? {description: model.description} : {}),
-    // By using Vega layout, we don't support custom autosize
-    autosize: topLevelProperties.autoResize ? {type: 'pad', resize: true} : 'pad',
+    autosize: autosize && isAutosize(autosize) ? {...{type: 'pad'}, ...autosize} : autosize || 'pad',
```

Comment:
I think we should just normalize autoSize to have an object form from the beginning, so we don't need complicated logic here nor  helper methods like `autosizeFit`?  (Just read `autosize.type === 'fit'`). 

(We mostly do the same thing for other similar structures with a short form such as mark and bin.)

Reply:
Unfortunately this doesn't quite work with the typings but I am adding normalization as it makes the code a bit easer.

Reply:
The trick is to declare 
```
const autoSize = normalizeAutoSize(topLevelProperties.autoSize); 
```
then for the rest of the code `autoSize` is `AutoSizeParams`

Reply:
I have a lot of places where I only have a spec. This works in one place only.

Reply:
I guess I can also do `normalizeAutoSize(spec.autosize).type === 'fit'`